lib/utils/pyexec.c

```diff
@@ -157,6 +157,8 @@ STATIC int parse_compile_execute(const void *source, mp_parse_input_kind_t input
         mp_hal_stdout_tx_strn("\x04", 1);
     }
 
+    //BOARD_HANDLE_EXC(nlr.ret_val, &ret);
```

Comment:
This was actually a TODO... I wanted to add a hook for a board to deal with any raised exception (eg clean up state, turn of things that shouldn't be running when you drop to the REPL) and here looked like a good place.  There's also possibly another place to add a hook in this file, at the start of the REPL loop (where it enables IRQs if they're disabled, and unlocks the GC if it's locked).

Reply:
This very useful. It might be a good idea to provide a somewhat more elaborate skeleton of code on how, for example, to proceed in case of a fatal error.

Reply:
I've now uncommented this and made it more generic, passing along `input_kind` and `exec_flags` so the function could detect the difference between running a file and running REPL code.

I also added a hook that runs before the execution.

Note that these hooks are not called before/after imported/exec/eval'd code, they're only called for `boot.py`, `main.py` and code executed at the REPL.  But I think that's actually what is needed for a board to intercept errors that propagate fully out (ie are never handled).

>  It might be a good idea to provide a somewhat more elaborate skeleton of code on how, for example, to proceed in case of a fatal error.

This hook here could modify `ret` to set the `PYEXEC_FORCED_EXIT` flag, which would cause the stm32 port to jump to the `soft_reset_exit` label.